### DISCUSSION TITLE
Obtaining Scalar Values for Postprocessing

### DISCUSSION THREAD

**Discussion by JKHINL** [idaholab/moose#18497] [Q&A Modules: General] [answered]
Hello,

I have generated a mesh using Cubit, and am currently in the process of writing an input file for Pronghorn (which uses MOOSE modules). The mesh mostly contains triangles (see Figure below). 

I have a list of instruments with positions in Cartesian coordinates, and would like to compare my simulation results (specifically, the variable T_solid) to the experimental data. The problem is that most of the locations of these instruments are not at nodes, nor situated on lines connecting two nodes. The imperfect solution would be to use NearestNodeNumber but it is giving me the following error:

*** ERROR ***
A 'NearestNodeNumberUO' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

I was wondering what the best approach to this would be, and which postprocessors (or userobjects) I should use, or what the issue is with using 'NearestNodeNumber'. 

Thanks,
Josh

![Core](https://user-images.githubusercontent.com/86485278/127932186-13f048ba-b41a-4850-9e3b-f6d8468488f5.PNG)

**Comment by WilkAndy** [comment]
This sort of question comes up so often: `blah is not a registered object`.   @JKHINL  , you can search the discussions for various answers.   Your problem is that [NearestNodeNumberUO](https://mooseframework.inl.gov/source/userobjects/NearestNodeNumberUO.html) is in the Geochemistry module, so you need to link that in.

i think we need to have something at mooseframework.org describing how to link parts of MOOSE and why you do/don't need to do it, written in simple English for people like me, since the Registry stuff is rather opaque.

**Reply by GiudGiud** [reply]
Should we move NearestNodeUO to the framework folder (since it's a more general capability) or just add a warning that this linking is needed to that object

**Reply by WilkAndy** [reply]
I'm happy for you to move it if you decide it's best in framework.

**Comment by aprilnovak** [comment, marked as answer]
Hi @JKHINL - another option could be to use the `PointValue` postprocessor, which will return the value of a variable at a point (I believe by evaluating the actual coefficients * basis functions used to represent that variable) - regardless of whether that point perfectly lines up with a node.

**Reply by JKHINL** [reply]
Hello @aprilnovak, I did try the PointValue postprocessor earlier this morning, but unfortunately it gave me a "No element located at (x,y,z)=(0.6, 0, 0) in PointValue Postprocessor" message, when that point does fall inside the geometric domain. Is there a way around this?

**Reply by aprilnovak** [reply]
Hmm, I'm not sure - I've never run into an issue like this myself. Maybe `PointValue` doesn't work if the point is on a boundary of the domain? (just a guess, since I presume `z = 0` might be on the boundary of your geometry).

**Reply by JKHINL** [reply]
In this case, the point (x,y,z)=(0.6, 0, 0) is situated right in the middle of the domain. The cylinder will stretch axially along the x-axis.

It will work, however, if it is at a node on a boundary (e.g. 0,0,0)

**Reply by GiudGiud** [reply]
Could you please try this one instead https://mooseframework.inl.gov/source/postprocessors/ElementalVariableValue.html
without having looked at the code, it feels like point value is not doing what we expect.

**Reply by GiudGiud** [reply]
@JKHINL is 0.6 0 0 on an edge? If so, as the documentation says, the behavior of that postprocessor is undefined if discontinuous variables are used. What variable type are you sampling?

**Reply by GiudGiud** [reply]
Raised an issue about it: #18509

**Reply by JKHINL** [reply]
@GiudGiud @aprilnovak I "reset" the coordinates of my mesh by translating it a distance and then back again in Cubit, and PointValue now works. I am not sure what the issue was but I have tested five difference points within the mesh (at nodes, on edges, etc) and they all appear to work.
